### ISSUE TITLE
`IsolationMode.InstancePerTest` not working for native tests

### ISSUE THREAD

**Issue by filipegeric** [kotest/kotest#3414]
**Kotest version: 5.5.1**

In a multiplatform project when running `nativeTest` `IsolationMode.InstancePerTest` is not working.
Setting it in the class doesn't work:
```kotlin
import io.kotest.core.spec.DisplayName
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe

@DisplayName("Running an example")
class ExampleTests : StringSpec() {
    var a = 0

    override fun isolationMode() = IsolationMode.InstancePerTest

    init {
        "should be 1 after 1 increment" {
            a++
            a shouldBe 1
        }

        "should be 2 after 2 increments" {
            a++
            a++
            a shouldBe 2
        }
    }

}
```
or setting it as a system property in `build.gradle.kts`:
```kotlin
tasks.withType<Test> {
    systemProperty("kotest.framework.isolation.mode", "InstancePerTest")
}
```

The result of running `./gradlew nativeTest` is the following (The error message is not descriptive):

```
> Task :nativeTest FAILED

ExampleTests.should be 2 after 2 increments FAILED
    io.kotest.assertions.AssertionFailedError at null:-1

2 tests completed, 1 failed

FAILURE: Build failed with an exception.
```

**Comment by kirillzh** [comment]
@Kantis, any chance this could be prioritized? 

We rely on isolated tests heavily and currently have to manually reset implementations of our fake dependencies in each `beforeTest`. We would love to apply isolation mode globally so that we can drop all of that bespoke configuration in favor of isolation mode, but currently blocked by it not working on iOS.

**Comment by Kantis** [comment]
Looks like this is simply not implemented.

The JVM executor will define a runner based on the isolation mode, as seen [here](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/createSpecExecutorDelegate.kt)

The native counterpart resides [here](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-engine/src/desktopMain/kotlin/io/kotest/engine/spec/createSpecExecutorDelegate.kt). 

Trying to move the SpecRunners to commonMain shows that they rely on some JVM-only classes, namely:
- `AtomicInteger` ([example](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/runners/InstancePerLeafSpecRunner.kt#L52))
- `ConcurrentHashMap` ([example](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/runners/InstancePerTestSpecRunner.kt#L70))
- `PriorityQueue` ([example](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/runners/InstancePerLeafSpecRunner.kt#L55))

I'm not very familiar with what's available on multiplatform. Is there anything we could use to replace these classes easily?

**Comment by kirillzh** [comment]
Thank you for context @Kantis, this is helpful!

- For atomic operations - https://github.com/Kotlin/kotlinx-atomicfu
- For concurrent mutable map - https://github.com/touchlab/Stately/blob/main/stately-concurrent-collections/src/commonMain/kotlin/co/touchlab/stately/collections/ConcurrentMutableMap.kt
- For `PriorityQueue`, as far as I know there's no official KMP implementation, but could probably be replaced with a custom implementation like [this](https://github.com/badoo/Reaktive/blob/c7c729b6b2355410a564dd96f8a43ae4c196d321/reaktive/src/commonMain/kotlin/com/badoo/reaktive/utils/queue/PriorityQueue.kt#L5)?

**Comment by sksamuel** [comment]
The blocker at the moment is that for native and JS tests, there is no constructor provided to Kotest. We capture a reference to an instance and invoke the tests on it.

In JVM land we have the class type and can use reflection to create new instances of it.

Until compiler plugins go stable (Kotlin 2.1? 2.2?) it will be too much of a lift to get this working in a way that's stable across versions (Jetbrains make breaking changes even on minor releases).

**Comment by kirillzh** [comment]
@sksamuel is there any chance that this will be resolved in Kotest 6.0 with Kotlin 2.0+ support?

**Comment by sksamuel** [comment]
I'm aiming to try and support it yes.

**Comment by kirillzh** [comment]
Tried with latest 6.0.0.M2 and still not seeing it work with K/N: https://github.com/kirillzh/kotest-playground/blob/isolation-mode-single-per-test/shared/src/commonTest/kotlin/com.jetbrains.basicsample/IsolationModeTests.kt. 

Apologies if it's already tracked elsewhere, or if it's expected not to work until 6.0 stable version yet.

**Comment by sksamuel** [comment]
Been a while but coming in 6.2 !
